docs/_ext/validating_code_block.py

```diff
@@ -0,0 +1,60 @@
+from typing import List
+from docutils import nodes
+from docutils.parsers.rst import Directive
+from docutils.parsers.rst import directives
+from sphinx.application import Sphinx
+from sphinx.util.docutils import SphinxDirective
+from sphinx.directives.code import CodeBlock
+from sphinx.errors import ExtensionError
+
+import os
+import subprocess
+
+
+class ValidatingCodeBlock(CodeBlock):
+  """A directive that provides protobuf yaml formatting and validation.
+
+    'type-name' option is required and expected to conain full Envoy API type.
+    An ExtensionError is raised on validation failure.
+    Validation will be skipped if SPHINX_SKIP_CONFIG_VALIDATION environment variable is set.
+    """
+  has_content = True
+  required_arguments = CodeBlock.required_arguments
+  optional_arguments = CodeBlock.optional_arguments
+  final_argument_whitespace = CodeBlock.final_argument_whitespace
+  option_spec = {
+      'type-name': directives.unchanged,
+  }
+  option_spec.update(CodeBlock.option_spec)
+  skip_validation = (os.getenv('SPHINX_SKIP_CONFIG_VALIDATION') or 'false').lower() == 'true'
+
+  def run(self):
+    source, line = self.state_machine.get_source_and_line(self.lineno)
+    # built-in directives.unchanged_required option validator produces a confusing error message
+    if self.options.get('type-name') == None:
+      raise ExtensionError("Expected type name in: {0} line: {1}".format(source, line))
+
+    if not ValidatingCodeBlock.skip_validation:
+      args = [
+          'bazel-bin/tools/config_validation/validate_fragment',
+          self.options.get('type-name'), '-s', '\n'.join(self.content)
+      ]
+      process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+      stdout, stderr = process.communicate()
```

Comment:
Nit: is there a reason not to just use `subprocess.check_call()` here?

Reply:
check_call/check_output seem to lose at least some of the output (and using PIPE with stderr isn't recommended). The process output is helpful as it points to the error in yaml.

Reply:
Have you tried `subprocess.run` (new in Python 3) at https://docs.python.org/3/library/subprocess.html? We're not a Python 3 only shop.